### PR TITLE
Issue 4391- Federation groups

### LINKED ISSUE #4391: The ability to specify a grouping for containers within a federation
### Description
At the moment a federation tree typically looks like this
````mermaid
graph TD
Root([Federation])
Container1([Container 1])
Container2([Container 2])
Container3([Container 3])
Container4([Container 4])
Root-->Container1
Root-->Container2
Root-->Container3
Root-->Container4
````

We want to be able to introduce groupings above the container level. e.g. the following tree contains 2 additional grouping, `Archiecture` and `Structure`:
````mermaid
graph TD
Root([Federation])
Architecture([Architecture])
Structure([Structure])
Container1([Container 1])
Container2([Container 2])
Container3([Container 3])
Container4([Container 4])
Root-->Architecture
Root-->Structure
Architecture-->Container1
Architecture-->Container2
Structure-->Container3
Root-->Container4

style Architecture fill:#ffff99,stroke:#e6e600
style Structure fill:#ffff99,stroke:#e6e600
````
#### Specification
- Containers doesn't have to be within a group (they can be connected to root of the federation as before
- There can be a maximum of 1 group level

#### Creating groups
They should be depicted and editable every time a federation is updated via the [new revision endpoint](https://staging.dev.3drepo.io/docs/#/Revisions/createNewFederationRevision)

at the moment we just take in a array of containers. as follows:
````js
{
  "containers: [
    "container 1 ID",
    "container 2 ID"
  ]
}
````

We should extend the schema to also take in a group field, i.e for the example above:
````js
{
  "containers: [
    { _id: "container 1", group: "Architecture"},
    { _id: "container 2", group: "Architecture"},
    { _id: "container 3", group: "Structure"},
    { _id: "container 4"}
  ]
}
````

This information should then be stored in model settings as they do now.

#### Visualising the group
The grouping needs to then be reflected on the JSON tree. The tree is generated dynamically at run time depending on the containers array.

### Goals
- [x] Be able to update a federation specifying the grouping
- [x] Be able to visualise the grouping within the tree

### Tasks
- [x] Store group information
  - [x] Update middleware to understand the new schema
  - [x] Store group schema in model settings subModels field 
  - [x] Ensure we still understand the old schema and convert it to the new one
- [x] Return group information
  - [x] cast old schema to new
  - [x] ensure new schema is understood by all callees
    - [x] v5
    - [x] v4
    - [x] scripts (if applicable)
- [x] Embed group information into the tree https://github.com/3drepo/3drepobouncer/issues/639
- [x] Migration script to remove old schema
- [x] Create new ticket for next release to remove conversion code https://github.com/3drepo/3drepo.io/issues/4427
- [x] Update swagger doc

### Test
-  v4
  - [ ] Can edit federations (pre/post migration)
  - [ ] Can create federation 
  - [ ] Can import BCF
  - [ ] account list is correct
- v5
  - [ ] Delete container - submodel check should function as before
  - [ ] Get TicketGroupByID should work as before (for purposes of getting ids from submodels)
  - [ ] GetModelStats should work
  - [ ] Can edit federations (pre/post migration)

### PR DESCRIPTION
**Pull request by carmenfan**
This fixes #4391

#### Description
New functionality added:
- The ability to assign groups to containers with a federation
  - schema has been altered to store this information in federation settings
  - swagger has been updated to reflect the new schema
- The ability to render the groups correctly in the tree
  - requires https://github.com/3drepo/3drepobouncer/issues/639
  - tree UI altered to understand where a model reference is in the tree (instead of assuming it's always the first child)

#### Test cases
- v4
  - [ ] Can edit federations (pre/post migration)
  - [ ] Can create federation 
  - [ ] Can import BCF
  - [ ] account list is correct
- v5
  - [ ] Delete container - submodel check should function as before
  - [ ] Get TicketGroupByID should work as before (for purposes of getting ids from submodels)
  - [ ] GetModelStats should work
  - [ ] Can edit federations (pre/post migration)

### CODE REVIEW

**Review by ChristosTsiotsias — CHANGES_REQUESTED** [review]

**Review by ChristosTsiotsias — APPROVED** [review]

File: frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederationContainersList/editFederationContainersListItem/groupOption/groupOption.component.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+				<Message>
+					<FormattedMessage id="tickets.groups.field.noOptions" defaultMessage="No options" />
+				</Message>
+			</Container>
+		);
+	}
+
+	if (!trimmedValue) return (<></>);
+
+	return (
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederationContainersList/editFederationContainersListItem/groupOption/groupOption.component.tsx:45]: can we add a bit of a comment here stating that the css styles and behaviours from materialuitem will be applied through props?

File: frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederation.component.tsx [hunk 2]
```diff
@@ -18,58 +18,48 @@
...
 	const includeContainer = (container: IContainer) => {
-		setIncludedContainers((existingState) => [...existingState, container]);
+		setIncludedContainers((oldIncludedContainers) => [...oldIncludedContainers, container]);
 	};
 
-	const removeContainer = (containerToRemove: IContainer) => {
-		setIncludedContainers((existingState) => existingState.filter((container) => container !== containerToRemove));
+	const removeContainer = ({ _id }: IContainer) => {
+		setIncludedContainers((oldIncludedContainers) => oldIncludedContainers.filter((container) => container._id !== _id));
 	};
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/federations/editFederationModal/editFederation/editFederation.component.tsx:54]: I think we could get rid of these two and just call  `includeContainers([container])` and  `removeContainers([container])`.

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — DISMISSED** [review]

**Pull request merged**